_episodes/02-basic-concepts.md

```diff
@@ -208,4 +233,239 @@ thread hierarchy and the meaning of each of the two arguments,
+
+While this code would work fawlessly for a single GPU device, it hides a potential multi-threading bug when more than a single GPU device
+is available. The reason is we used OpenMP to spawn more threads without specifying the device. CUDA assumes device 0 is the default device, which 
+might lead to unexpected behavior. To fix this problem, we need to use the API function: `cudaError_t cudaSetDevice (int  device)` to specify
+which device to run multi-processing on. The improved code is shown below.
+
+~~~
+#include <stdio.h>                        /* For printf() function */
+
+/**********************************************/
+
+
+int main(int argc, char **argv) {
+
+    int device = atoi(sys.argv[1]), numGPUs;
+
+    checkCudaErrors(cudaGetDeviceCount(&numGPUs));
+
+    printf("CUDA-capable device count: %i\n", numGPUs);
+        
+    helloFromCPU();                       /* Calling from host */
+
+    #pragma omp parallel 
+    {
+        checkCudaErrors(cudaSetDevice(device));    /* Run OpenMP only on device 1 */
+        helloFromGPU<<<1, 8>>>();             /* Calling from device */
+    }
+
+    checkCudaErrors( cudaDeviceSynchronize() );    /* Force the host to wait for the device to check for error */ 
+
+    /* Do more stuff ... */
+
+    checkCudaErrors( cudaDeviceReset() );         /* House-keeping on device */
+
+    return(EXIT_SUCCESS);
+}
+~~~
+{: .language-cuda}
+
+Compiling the new ***hello.cu*** code via 
+
+~~~
+$ nvcc hello.cu -Xcompiler -fopenmp -o hello
+~~~
+{: .language-bash}
+
+Running the executable on device 0 works without any problems:
+
+~~~
+$ ./hello 0
+~~~
+{: .language-bash}
+
+will print "Hello World from GPU!" `nCPUcores` times:
+
+~~~
+Hello World from CPU!
+Hello World from GPU!
+...
+Hello World from GPU!
+~~~
+{: .output}
+
+However, running the executable on device 1 should return error if only a single CUDA-capable GPU device is available.
+
+~~~
+$ ./hello 1
+~~~
+{: .language-bash}
+
+will print the following output:
+
+~~~
+CUDA-capable device count: 1
+Hello World from CPU!
+CUDA error at test.cu:37 code=101(cudaErrorInvalidDevice) "cudaSetDevice(1)" 
+...
+~~~
+{: .output}
+
+If we remove the error checking function (checkCudaErrors) before cudaSetDevice(1) and recompile, the exectuable runs successfully, so we couldn't
+have known in runtime that there's something flawed with our code. As you can see, finding such bugs can be quite difficult without error checking. This is because in principle, we can launch multiple kernels that run on device 0 but access memory allocated on device 1. This could cause invalid memory access or corruption errors if we're using pointers to allocate and access memory. Therefore, you should always make it a habit to set the device whenever your 
+code could potentially spawn new host threads, and check for runtime errors every single time you make an API function call.
+
+> ## Note:
+> Always set the GPU device when creating new host threads
+{: .discussion}
+
```

Comment:
Multiple GPUs are a topic for advanced level. Take a look at our plans for GPU programming tutorial within Education project in Asana :See https://app.asana.com/0/1143574976003361/1199970266470673/f

Reply:
You might be right but I still think checking for multiple devices is important (even for beginners), I agree the openMP part coupled to GPUs might be more suitable for intermediate-advanced level.

Reply:
They have separate courses for multiple GPUs at Nvidia as well which are labled intermediate. 
I think it would be more appropriate to move them to intermediate level from this PR. Then we can work to improve them in place.